state/machine.go

```diff
@@ -602,6 +602,48 @@ func (m *Machine) PasswordValid(password string) bool {
 // a HasAssignedUnitsError.  If the machine has containers, Destroy
 // will return HasContainersError.
 func (m *Machine) Destroy() error {
+	if m.IsManager() {
+		units, err := m.Units()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		unitsMap := make(map[string]*Unit)
+		for _, unit := range units {
+			unitsMap[unit.String()] = unit
+		}
+		var evacuablePrincipals []string
+		for _, principalUnit := range m.doc.Principals {
+			unit, ok := unitsMap[principalUnit]
+			if !ok {
+				continue
+			}
+			curl, err := unit.CharmURL()
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if strings.HasPrefix(curl.Name, "juju-") {
```

Comment:
I wonder if we just want to restrict this to units that the user is not responsible for. Right now, that's just controller charm. They need to manually deploy `juju-dashboard` charm so I am not sure we should count that.

Reply:
true, but my feeling here is that it should be more generic